common/str-enc.cpp

```diff
@@ -20,94 +20,567 @@
+				chr |= (src[i++] & 0x3F) << 6;
+				chr |= (src[i++] & 0x3F);
 				break;
 
 			case 3:
-				chr |= (_str[i++] & 0x0F) << 12;
-				chr |= (_str[i++] & 0x3F) << 6;
-				chr |= (_str[i++] & 0x3F);
+				chr |= (src[i++] & 0x0F) << 12;
+				chr |= (src[i++] & 0x3F) << 6;
+				chr |= (src[i++] & 0x3F);
 				break;
 
 			case 2:
-				chr |= (_str[i++] & 0x1F) << 6;
-				chr |= (_str[i++] & 0x3F);
+				chr |= (src[i++] & 0x1F) << 6;
+				chr |= (src[i++] & 0x3F);
 				break;
 
 			default:
-				chr = (_str[i++] & 0x7F);
+				chr = (src[i++] & 0x7F);
 				break;
 			}
 		} else {
 			break;
 		}
 
-		dst += chr;
+		operator+=(chr);
+	}
+}
+
+const uint16 invalidCode = 0xFFFD;
+
+static bool cjk_tables_loaded = false;
+static const uint16 *windows932ConversionTable;
+static const uint16 *windows949ConversionTable;
+static const uint16 *windows950ConversionTable;
+
+static const uint16 *loadCJKTable(File &f, int idx, size_t sz) {
+	f.seek(16 + idx * 4);
+	uint32 off = f.readUint32LE();
+	f.seek(off);
+	uint16 *res = new uint16[sz];
+	f.read(res, 2 * sz);
+#ifndef SCUMM_LITTLE_ENDIAN
+	for (uint i = 0; i < sz; i++)
+		res[i] = FROM_LE_16(res[i]);
+#endif
+	return res;
+}
+
+static void loadCJKTables() {
+	File f;
+
+	cjk_tables_loaded = true;
+
+	if (!f.open("encoding.dat")) {
+		warning("encoding.dat is not found. Support for CJK is disabled");
+		return;
+	}
+
+	if (f.size() < 16 + 3 * 4) {
+		warning("encoding.dat is invalid. Support for CJK is disabled");
+		return;
+	}
+
+	if (f.readUint32BE() != MKTAG('S', 'C', 'V', 'M')
+	    || f.readUint32BE() != MKTAG('E', 'N', 'C', 'D')) {
+		warning("encoding.dat is invalid. Support for CJK is disabled");
+		return;
+	}
+
+	// Version and number of tables.
+	if (f.readUint32LE() != 0 || f.readUint32LE() < 3) {
+		warning("encoding.dat is of incompatible version. Support for CJK is disabled");
+		return;
+	}
+
+	windows932ConversionTable = loadCJKTable(f, 0, 47 * 192);
+	windows949ConversionTable = loadCJKTable(f, 1, 0x7e * 0xb2);
+	windows950ConversionTable = loadCJKTable(f, 2, 89 * 157);
+}
+
+void U32String::decodeWindows932(const char *src, uint32 len) {
+	ensureCapacity(len, false);
+
+	if (!cjk_tables_loaded)
+		loadCJKTables();
+
+	for (uint i = 0; i < len;) {
+		uint8 high = src[i++];
+
+		if ((high & 0x80) == 0x00) {
+			operator+=(high);
+			continue;
+		}
+
+		// Katakana
+		if (high >= 0xa1 && high <= 0xdf) {
+			operator+=(high - 0xa1 + 0xFF61);
+			continue;
+		}
+
+		if (i >= len) {
+			operator+=(invalidCode);
+			continue;
+		}
+
+		uint8 low = src[i++];
+		if (low < 0x40) {
+			operator+=(invalidCode);
+			continue;
+		}
+		uint8 lowidx = low - 0x40;
+		uint8 highidx;
+
+		if (high >= 0x81 && high <= 0x84)
+			highidx = high - 0x81;
+		else if (high >= 0x87 && high <= 0x9f)
+			highidx = high - 0x87 + 4;
+		else if (high >= 0xe0 && high <= 0xee)
+			highidx = high - 0xe0 + 29;
+		else {
+			operator+=(invalidCode);
+			continue;
+		}
+
+		if (!windows932ConversionTable) {
```

Comment:
This test is here I assumed for the case where we failed to load the table.
But where is the `windows932ConversionTable` set to `nullptr`? From a quick look at your code it seems that if we fail to load the tables, the table pointers are left uninitialised?

Reply:
They're static, so they're implicitly inited to 0

Reply:
They're static, hence inited to null implicitly